### PR TITLE
tickets #1510 and #1511

### PR DESCRIPTION
**Pull request by Voyrx**
Pull request related to https://github.com/privacyidea/privacyidea/issues/1510

### CODE REVIEW

**Comment by cornelinux** [comment]
Thanks a lot for the pull request, but I think we should split this in several PRs.

Please provide a PR for the fix of the certificate directory in the localca. This will be easy to merge.

I will add some further questions in your code.

File: privacyidea/lib/eventhandler/tokenhandler.py [hunk 1]
```diff
@@ -391,6 +391,9 @@ def do(self, action, options=None):
                                         "has no email address!".format(user))
                 elif tokentype == "motp":
                     init_param['motppin'] = handler_options.get("motppin")
+                    
+                elif tokentype == "certificate":
+                    init_param['request'] = request.all_data.get("request")
```
Comment (cornelinux) [inline privacyidea/lib/eventhandler/tokenhandler.py]: What is your intention with enrolling a certificate.
Where does actually the "request" parameter come from, which is usually not contained in the originial http request.

The problem of this approach is, that the originial https request needs to know, that there is an event handler configured, that needs an additional parameter. 
We should think about a more generic way to solve this.

File: privacyidea/lib/tokens/certificatetoken.py [hunk 2]
```diff
@@ -334,3 +339,79 @@ def revoke(self):
...
+    
+    @staticmethod
+    def api_endpoint(request, g):
+        """
+        Author : Tarek EL ALLAM
+        This provides a function to be plugged into the API endpoint
+        /ttype/<tokentype> which is defined in api/ttype.py
+        See :ref:`rest_ttype`.
+        Desc : 
+        Add API call in order to get the last generated certificate assigned, activated, unrevoked and unlocked.
```
Comment (cornelinux) [inline privacyidea/lib/tokens/certificatetoken.py]: actually we already have such an endpoint, it is the ``GET /token/`` endpoint, which needs an administrator or a user to authenticate. But you actually can have the same result.
This approach leaves the information, which certificate the user has, open to the world without any authentication.
Why would you do this?
Would you be able to have a service account to receive this information?

**Comment by Voyrx** [comment]
Hello,

The "request" corresponds to the certificate signing request, I used it like that because I noticed that it's a native parameter for GUI certificate enrollement :
[cid:2956fb3e-b50b-44e3-922c-50c1766391ac]

But of course, if you think that it's not beautiful, I can adapt the solution. 😄

Thank you for you quick answer.

Regards,
Tarek EL ALLAM

________________________________
De : Cornelius Kölbel <notifications@github.com>
Envoyé : mardi 19 mars 2019 23:12
À : privacyidea/privacyidea
Cc : Tarek EL ALLAM; Author
Objet : Re: [privacyidea/privacyidea] tickets #1510 and #1511 (#1512)


@cornelinux commented on this pull request.

________________________________

In privacyidea/lib/eventhandler/tokenhandler.py<https://github.com/privacyidea/privacyidea/pull/1512#discussion_r267119080>:

> @@ -391,6 +391,9 @@ def do(self, action, options=None):
                                         "has no email address!".format(user))
                 elif tokentype == "motp":
                     init_param['motppin'] = handler_options.get("motppin")
+
+                elif tokentype == "certificate":
+                    init_param['request'] = request.all_data.get("request")


What is your intention with enrolling a certificate.
Where does actually the "request" parameter come from, which is usually not contained in the originial http request.

The problem of this approach is, that the originial https request needs to know, that there is an event handler configured, that needs an additional parameter.
We should think about a more generic way to solve this.

**Pull request closed without merging**